### PR TITLE
feat: Unify CSS and JS breakpoints utilities

### PR DESCRIPTION
**Pull request by aaronccasanova**
### WHAT is this pull request doing?

This PR unifies our breakpoint utilities in CSS and JS and exposes our SCSS media conditions from `@shopify/polaris-tokens`:
- Created `getMediaConditions` to transform `breakpoints` tokens into a generic structure for creating CSS and JS utilities
- Added a `toMediaConditions` transform to `@shopify/polaris-tokens` for exposing SCSS media conditions from `breakpoints` tokens (Needed for upcoming Web breakpoint upgrades)
- Added a `useBreakpoints` hook to `@shopify/polaris` exposing the same media conditions available in the SCSS API
- Removed `polaris-react/scripts/gen-media-queries` and associated npm run-scripts
- Updated `polaris-react/src/styles/common` to use `@shopify/polaris-tokens/dist/scss/media-queries.scss`

`getMediaConditions` example:

```js
const mediaConditions = getMediaConditions(tokens.breakpoints)
/*
mediaConditions === {
  'breakpoints-xs': {
    up: '(min-width: ...)',
    down: '(max-width: ...)',
    only: '(min-width: ...) and (max-width: ...)',
  },
  'breakpoints-sm: {...},
  'breakpoints-md: {...},
  //etc...  
}
*/
```

`@shopify/polaris-tokens` `media-queries.scss` example:

```scss
$p-breakpoints-xs-up: '(min-width: ...)';
$p-breakpoints-xs-up: '(max-width: ...)';
$p-breakpoints-xs-up: '(min-width: ...) and (max-width: ...)';

$p-breakpoints-sm-up: '(min-width: ...)';
$p-breakpoints-md-up: '(min-width: ...)';
// etc...
```

`@shopify/polaris` `useBreakpoints` example:

```js
const {
  xsUp,
  xsDown,
  xsOnly,
  smUp,
  mdUp,
  // etc...
 } = useBreakpoints()
```
<details>
  <summary><code>useBreakpoints()</code> demo</summary>
  
Notice the transition from `smOnly` to `mdUp` and `mdDown` to `mdOnly`
  
https://user-images.githubusercontent.com/32409546/177426715-e861ad5d-482d-482a-bb58-a2aa17b8c5b7.mov

</details>

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-tokens/package.json [hunk 1]
```diff
@@ -27,7 +27,8 @@
   },
   "devDependencies": {
     "@shopify/browserslist-config": "^3.0.0",
-    "ts-node": "^10.7.0"
+    "ts-node": "^10.7.0",
+    "type-fest": "^2.13.1"
```
Comment (BPScott) [inline polaris-tokens/package.json]: The contents of the one type that you need here looks pretty trivial: https://unpkg.com/browse/type-fest@2.14.0/source/exact.d.ts

What do ya reckon to copying it into polaris-tokens?


```
type Primitive = null | undefined | string | number | boolean | symbol | bigint;
type KeysOfUnion<T> = T extends T ? keyof T : never;

export type Exact<ParameterType, InputType extends ParameterType> = ParameterType extends Primitive
	? ParameterType
	: {[Key in keyof ParameterType]: Exact<ParameterType[Key], InputType[Key]>} & Record<Exclude<keyof InputType, KeysOfUnion<ParameterType>>, never>;
```

Comment (aaronccasanova) [reply polaris-tokens/package.json]: I replaced the copy/paste types with the intention of using more `type-fest` utilities, but only ended up needing `Entry<...>`. [Reverted the dev dep here](https://github.com/Shopify/polaris/pull/6407/commits/7934d32f617fb9a688b58e8eb9458dba3c31fc29) and introduced a naive `Entry<...>` for now.

File: polaris-react/playground/Playground.tsx [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
 import {Page} from '../src';
+import {useBreakpoints} from '../src/utilities/breakpoints';
 
 export function Playground() {
+  const breakpoints = useBreakpoints();
+
   return (
     <Page title="Playground">
-      {/* Add the code you want to test in here */}
+      {Object.entries(breakpoints).map(([name, value]) => (
```
Comment (sam-b-rose) [inline polaris-react/playground/Playground.tsx]: Nice example!

Be sure to remove this from `Playground.tsx` before merging 👍

File: polaris-react/src/styles/_common.scss [hunk 3]
```diff
@@ -18,4 +18,4 @@
 @import './shared/typography';
 @import './shared/interaction-state';
 
-@import './media-queries';
+@import '../../../node_modules/@shopify/polaris-tokens/dist/scss/media-queries';
```
Comment (sam-b-rose) [inline polaris-react/src/styles/_common.scss:21]: Wish there was a better way of resolving from `node_modules`. Does using the `~` prefix work for this?

Example: `@import '~@shopify/polaris-tokens/dist/scss/media-queries';`

File: polaris-tokens/scripts/index.ts [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
 
 import {toJSON} from './toJSON';
+import {toMediaConditions} from './toMediaConditions';
 import {toStyleSheet} from './toStyleSheet';
 
 (async () => {
-  await Promise.all([toJSON(tokens), toStyleSheet(tokens, osColorSchemes)]);
+  await Promise.all([
+    toJSON(tokens),
+    toMediaConditions(tokens.breakpoints),
```
Comment (sam-b-rose) [inline polaris-tokens/scripts/index.ts:10]: Really like how this fits in well with how we are generating other assets!

File: polaris-tokens/src/utilities.ts [hunk 5]
```diff
@@ -121,3 +122,87 @@ export function getCustomPropertyNames(tokens: Tokens) {
...
+
+export type BreakpointsTokenName = keyof BreakpointsTokenGroup;
+
+/**
+ * Alias extracted from each Polaris `breakpoints` token name.
+ *
+ * @example 'xs' | 'sm' | 'md' | 'lg' | 'xl'
+ */
+export type BreakpointsAlias =
+  BreakpointsTokenName extends `${string}-${infer Alias}` ? Alias : never;
```
Comment (sam-b-rose) [inline polaris-tokens/src/utilities.ts:136]: typescript fanciness ✨  today I learned about `infer`

**Review by gui-santos — APPROVED** [review]
🔥

**Review by sam-b-rose — APPROVED** [review]
Very nice, lots of awesome work here! I'd say you deserve a `break` after all these improvements and features 💯

File: polaris-react/src/styles/_common.scss (hunk 3 above)
Comment (aaronccasanova) [reply polaris-react/src/styles/_common.scss:21]: Totally! We briefly looked into this during the `CustomProperties` / global styles integration and it requires some infra updates. There's only two instances of importing styles from `node_modules` so I'm not sure the juice is worth the squeeze.

File: polaris-react/playground/Playground.tsx (hunk 2 above)
Comment (aaronccasanova) [reply polaris-react/playground/Playground.tsx]: https://github.com/Shopify/polaris/pull/6407/commits/f4f67fd14d4cbcd2a665e8e477ba17c3db0eb0f4

**Pull request merged**